### ISSUE TITLE
Singularity ignores SINGULARITYENV_ for variables defined in host ENV

### ISSUE THREAD

**Issue by zyndagj** [apptainer/singularity#4827]
### Version of Singularity:

```
$ singularity version
3.5.1-rc.2
```

### Expected behavior

- We expect `SINGULARITYENV_` to take precedence over host environment variables, so those values can be changed without affecting the host stability. While the `-e` flag does offer a way forward, removes ALL host environment variables which could be necessary for execution.
- We expect environments to be inherited the same between docker and singularity sources

#### 1. When environment variables are defined in a container recipe
  - singularity - variables can never be overridden
  - docker - variables can be overridden with
    - export VAR=new; singularity exec
    - `SINGULARITYENV_VAR=new singularity exec`
    - `export SINGULARITYENV_VAR=new; singularity exec`
       - **NOTE** - `SINGULARITYENV_VAR` is ignored if `VAR` is exported to the host ENV. The `-e` flag can ignore the host env, allowing `SINGULARITYENV_` to work, but not ideal for making small changes when host env contains things like filesystem paths.

#### 2. When environment variables are not defined in container recipe, but exist in host ENV
  - singularity - host variables can be overridden by
     - `export VAR=new; singularity exec`
     - `export SINGULARITYENV_VAR=new; singularity exec -e`
     - `SINGULARITYENV_VAR=new singularity exec -e`
  - docker - host variables can be overridden by
     - `export VAR=new; singularity exec`
     - `export SINGULARITYENV_VAR=new; singularity exec -e`
     - `SINGULARITYENV_VAR=new singularity exec -e`

### Steps to reproduce this behavior

1. Create `example.sif`
```
$ cat << 'EOF' > example.def
BootStrap: docker
From: ubuntu:xenial

%environment
    export LC_ALL=C
    export INTERNAL=1
EOF

singularity build example.sif example.def
```

2. Pull docker example
```
singularity pull docker://gzynda/sing_test:latest
```

3. Create `test.sh` script
```
#!/bin/bash

#!/bin/bash

function sing {
        singularity exec example.sif bash -c 'echo "Singularity |    | Internal: $INTERNAL External: $EXTERNAL"'
        singularity exec -e example.sif bash -c 'echo "Singularity | -e | Internal: $INTERNAL External: $EXTERNAL"'
}
function docker {
        singularity exec sing_test_latest.sif bash -c 'echo "Docker      |    | Internal: $INTERNAL External: $EXTERNAL"'
        singularity exec -e sing_test_latest.sif bash -c 'echo "Docker      | -e | Internal: $INTERNAL External: $EXTERNAL"'
}

export EXTERNAL=1

echo -e "EXTERNAL=1\nInjected variables"
SINGULARITYENV_INTERNAL=3 SINGULARITYENV_EXTERNAL=3 sing
SINGULARITYENV_INTERNAL=3 SINGULARITYENV_EXTERNAL=3 docker

echo "Defined variables"
SINGULARITYENV_INTERNAL=3
SINGULARITYENV_EXTERNAL=3
sing; docker

echo "Exported variables"
export SINGULARITYENV_INTERNAL=3
export SINGULARITYENV_EXTERNAL=3
sing; docker

export INTERNAL=2
unset SINGULARITYENV_INTERNAL SINGULARITYENV_EXTERNAL

echo -e "\nEXTERNAL=1 INTERNAL=2\nInjected variables"
SINGULARITYENV_INTERNAL=3 SINGULARITYENV_EXTERNAL=3 sing
SINGULARITYENV_INTERNAL=3 SINGULARITYENV_EXTERNAL=3 docker

echo "Defined variables"
SINGULARITYENV_INTERNAL=3
SINGULARITYENV_EXTERNAL=3
sing; docker

echo "Exported variables"
export SINGULARITYENV_INTERNAL=3
export SINGULARITYENV_EXTERNAL=3
sing; docker
```
Which yields
```
$ bash test.sh 
EXTERNAL=1
Injected variables
Singularity |    | Internal: 1 External: 1
Singularity | -e | Internal: 1 External: 3
Docker      |    | Internal: 3 External: 1
Docker      | -e | Internal: 3 External: 3
Defined variables
Singularity |    | Internal: 1 External: 1
Singularity | -e | Internal: 1 External: 
Docker      |    | Internal: 1 External: 1
Docker      | -e | Internal: 1 External: 
Exported variables
Singularity |    | Internal: 1 External: 1
Singularity | -e | Internal: 1 External: 3
Docker      |    | Internal: 3 External: 1
Docker      | -e | Internal: 3 External: 3

EXTERNAL=1 INTERNAL=2
Injected variables
Singularity |    | Internal: 1 External: 1
Singularity | -e | Internal: 1 External: 3
Docker      |    | Internal: 2 External: 1
Docker      | -e | Internal: 3 External: 3
Defined variables
Singularity |    | Internal: 1 External: 1
Singularity | -e | Internal: 1 External: 
Docker      |    | Internal: 2 External: 1
Docker      | -e | Internal: 1 External: 
Exported variables
Singularity |    | Internal: 1 External: 1
Singularity | -e | Internal: 1 External: 3
Docker      |    | Internal: 2 External: 1
Docker      | -e | Internal: 3 External: 3
```

### What OS/distro are you running

```
$ cat /etc/os-release

NAME="Ubuntu"
VERSION="16.04.6 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.6 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```

### How did you install Singularity

Compiling and installing from source.

**Comment by dtrudg** [comment]
Hi @zyndagj. Agreed that this inconsistency isn't great here. We've arrived here after a lot of changes over versions of Singularity (2.x and 3.x) to exactly how environment variables are handled. There have been a bunch of previous requests around particularly docker container env var handling, and they've sometimes been conflicting requests.

My feeling is that although a precedence change may look simple here, we shouldn't make a behavior change in the 3.5 series. For a future 3.6 we need to go back and look at all the requests people have had about env var handling, en masse, to identify the most sensible consistent situation.

**Comment by zyndagj** [comment]
That sounds reasonable. I didn't expect a change like this to squeeze in during the testing period of 3.5.1, I just wanted to make sure the behavior I saw on our systems was still present in your latest release.

I would however advocate for updating your 3.5 documentation on [Environment and Metadata](https://sylabs.io/guides/3.4/user-guide/environment_and_metadata.html?highlight=environment) to at least acknowledge these edge cases, so fewer tickets like mine get submitted.

**Comment by andyneff** [comment]
For anyone using Singularity 3.0 through 3.5 and doesn't have the authority to upgrade singularity on the system to use 3.6 or newer: I just ran into the issue, but it turns out a few details can make this issue even more unpredictable

The "order" of environment variables on the host affects how the variables are set when creating the singularity environment. Specifically, in these versions of singularity (which, I should reiterate, has been fixed in singularity 3.6), when processing `VAR` vs `SINGULARITYENV_VAR`, which ever comes last in the [environment variable list](https://golang.org/pkg/os/#Environ) wins.

Using [this gist](https://gist.github.com/andyneff/9e02194ee5ab4d5edcc88f332ae4be0e), I was able to see that:

- In bash, exporting a variable (E.g. `export foo=bar; singularity ...`) puts it into a list with "some" order (maybe hashed order?) and this changes from version to version of bash. What this means is that depending on the name of `VAR`, it could come before or after the `SINGULARITYENV_VAR` counterpart. Consequently, it becomes difficult to predict which method wins; i.e., which value the variable will take in the singularity container

- Using parameter assignment (e.g. `foo=bar singularity ...`), parameter assigned environment variables are always inserted first, before exported variables in the `environ` list, but are, themselves, also in "some" order.

The take away I got from this is if I'm going to use `SINGULARITYENV_` in 3.0-3.5, I should unset the non-SINGULARITYENV_ versions of the variables so that they won't interfere
